### PR TITLE
Lot tags feature

### LINKED ISSUE #2662: feat(stock): Adding a tag system for stock entry
This issue is about to have a mean of tagging the stock with any key words (project name).

This feature will be useful for knowing in a given depot, the stock which correspond to one or more tags.

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR add tagging feature in lots registry

A tag module exists in /Administration/Tags : we have to create tags first then we can use those tags in the lots registry

Capture d’écran 2020-07-28 à 22 08 20
Capture d’écran 2020-07-28 à 22 08 29
Capture d’écran 2020-07-28 à 22 08 48
Capture d’écran 2020-07-28 à 22 08 59

Close #2662

### CODE REVIEW

**Comment by jmcameron** [comment]
How will the tags be used?

- I see the Admin/Tags page that allows the creation of arbitrary tags.  I assume that tags will be used (eg, in Lots management) to facilitate searching.  However, it is confusing to me why we need a global pool of tags.   I suppose that is to prevent mistyping.   But that makes it difficult to create new tags when we are in the Lots Management page.   We may need a popup menu on the Lots page (and others as gets used in other pages) that will allow creating tags on the fly (rather than leaving and going to the Admin/Tags page).   
- A related question is why should we share the same pool of tags with everything.  Shouldn't the lots tags be distinct from all the other modules that need tags?  This could be done "under the hood" by associating a "module/user" with each tag (selected from a fixed list).  Then anytime we see tags in the Lots module, it would selected from the right kind of tags.   There are other ways to handle this.  Not sure it is worth the effort.   
- In the Admin/Tags management, is there any way to find out where specific tags are being used?   For instance, if we want to delete or rename a tag, it would be useful to know if that might cause problems by the items using that tag.

**Comment by mbayopanda** [comment]
Thank you @jmcameron for your review : 

- The concept of tags is not limited to lots only, the idea is we can use tags anywhere I know that the Tags Module is too minimal but we can improve it

- I really like your idea to know where a tag is used, actually, it is not implemented but we can do that

**Comment by mbayopanda** [comment]
@jmcameron I am open to all suggestions that you can make 👍

**Comment by jmcameron** [comment]
If this PR is intended to resolve Issue https://github.com/IMA-WorldHealth/bhima/issues/2662, please link it.  If it is to partially address the issue, please mention this issue in the description at the top.

**Comment by mbayopanda** [comment]
@jmcameron I have a question about how can we link the tags to where they are used? because the idea behind tags is we can use them not just with lots only, maybe it can require a new module or registry

**Comment by jniles** [comment]
@mbayopanda what is the progress here?  When will you be ready for another review?

**Comment by mbayopanda** [comment]
@jniles can I get a review, I skipped some tests which are now broken

**Review by jmcameron — APPROVED** [review]
I skimmed through the code and did not spot any problems.  I tried testing the basic functionality using bhima_test and it seems to work okay.    So it is ready to go, as far as I'm concerned but I do suggest you get one more review from someone else on the team.

**Comment by jmcameron** [comment]
> I skipped some tests which are now broken

What does this mean?  It looks like there are some good tests you've added.  Do feel any additional tests need to be added?

**Comment by jmcameron** [comment]
I would still like to get an explanation of how tags are going to be used.   I think this addition is fine (I did approve it), but it is not clear to me how the users of BHIMA will use this feature.

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [inline client/src/js/components/bhTagSelect/bhTagSelect.js:7]: Are these optional?  If so, can you make it `<?`?

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [inline client/src/js/components/bhTagSelect/bhTagSelect.js:24]: Why is this needed?  I don't see and `$broadcast` or `$emit` calls.

Generally, we should probably avoid the event emitter pattern.  We've used it a couple times (for example, I used it in the cash modal), but those are usually when we didn't know how to make a proper data model passing data back and forth.  In general, we should use bindings for inputs and callbacks for outputs.

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [inline client/src/js/components/bhTagSelect/bhTagSelect.js]: Why `read(null)`?

**Comment by mbayopanda** [comment]
@jmcameron we can think about tag usage as for example if we have computers in our depot, and we want to separate computer destined to each service of the organization, this case can occur if in the depot all computers for all purchase orders but we can have some computers purchased just for a given service and we don't want to give them to other services.

Tags can help us to perform this kind of distinction

**Comment by jmcameron** [comment]
> Tags can help us to perform this kind of distinction

Thank you @mbayopanda.  That was helpful.

File: client/src/i18n/en/form.json
Comment (jniles) [inline client/src/i18n/en/form.json]: I think "select tags to add" is a more natural way of saying this in English.

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [inline client/src/js/components/bhTagSelect/bhTagSelect.js]: Is `t` ever `null`?

Comment (jniles) [reply client/src/js/components/bhTagSelect/bhTagSelect.js]: Also, can you just put 'font-size' as a `style` on the element?

File: client/src/modules/debtors/groups.update.html
Comment (jniles) [inline client/src/modules/debtors/groups.update.html:115]: 👍

File: client/src/modules/stock/lots/registry.js
Comment (jniles) [inline client/src/modules/stock/lots/registry.js]: Since this is in `for loop` in a cell in a `uiGrid`, for performance reasons, let's pre-process this.  Stick it right on the tag like this:
```js
vm.gridOptions.data = lots.map(lot => {
  let tagNames = []
  lot.tags.forEach(tag => {
    tag.style = getTagColor(t);
  });
   return lot;
});
```

Now, in your cell, you can use `ng-style="tag.style"`.  That should make it not have as many $digest calls.

Comment (jniles) [reply client/src/modules/stock/lots/registry.js]: Also, does the `font-size` do anything here?

File: client/src/modules/tags/modal/createUpdate.js
Comment (jniles) [inline client/src/modules/tags/modal/createUpdate.js]: Since this is more than a `iconColor` (it has the font size in it), it might make sense to call it `iconStyle`.

File: client/src/modules/tags/modal/createUpdate.js
Comment (jniles) [inline client/src/modules/tags/modal/createUpdate.js]: It might be cleaner to just turn this into an array.

```js
// delete all iconColors on all tags
[].concat(vm.tags).forEach(tag => {
  delete tag.iconColor;
});
```

File: client/src/modules/tags/tags.ctrl.js
Comment (jniles) [inline client/src/modules/tags/tags.ctrl.js]: Since you do this about 3+ times in this PR, I suggest you put this in a Service.  Like the `TagService`.  Or `ColorService`. 

```js
Tags.formatTagStyle = tag => {
  tag.style = { color : tag.color, 'font-size': '14px' };
};
```

File: client/src/modules/tags/templates/name.cell.html
Comment (jniles) [inline client/src/modules/tags/templates/name.cell.html]: Do you need the `data-row` binding?  I can't tell where it is used.

File: client/src/modules/tags/tags.service.js
Comment (jniles) [inline client/src/modules/tags/tags.service.js:20]: Ooooh.  I had forgotten we had this service.  Excellent 👍

File: client/src/modules/tags/tags.service.js
Comment (jniles) [inline client/src/modules/tags/tags.service.js:26]: Beautiful 👍

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js:526]: 👍

File: test/end-to-end/tags/tags.spec.js
Comment (jniles) [inline test/end-to-end/tags/tags.spec.js]: Do you know why these fail?  Any ideas?

File: server/models/schema.sql
Comment (jniles) [inline server/models/schema.sql]: Since you are adding this property, I would go ahead and make it much longer.  VARCHAR(20) is super limiting.

**Review by jniles — COMMENTED** [review]
@mbayopanda I haven't given this a proper test, but I've looked over the code.  It looks good!  I'm a bit concerned about the performance implications of these tags, especially in the JOIN and the ui-grid.  I've added a few comments to try and improve on the client-side grid performance and tidy up the code.

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (mbayopanda) [reply client/src/js/components/bhTagSelect/bhTagSelect.js:24]: The reason of that is when we call the `tag modal` via the `lot edit modal` and we add a new `tag` from this `lot edit modal`, we must add this listener if not the list of tags loaded in the `lot edit modal` will not be up to date

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (mbayopanda) [reply client/src/js/components/bhTagSelect/bhTagSelect.js:7]: No, they were required when I created that component

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (mbayopanda) [reply client/src/js/components/bhTagSelect/bhTagSelect.js]: yes `t` can be `null` if nothing is selected from the list of tags

File: client/src/modules/tags/templates/name.cell.html
Comment (mbayopanda) [reply client/src/modules/tags/templates/name.cell.html]: It was used for `GridRow` tool for e2e tests, it is used also with: `/modules/roles/templates/action.cell.html`

File: test/end-to-end/tags/tags.spec.js
Comment (mbayopanda) [reply test/end-to-end/tags/tags.spec.js]: Just after adding the color picker in the tag modal, these tests started to fail

**Comment by mbayopanda** [comment]
Thank you @jniles for the review

File: client/src/modules/tags/tags.ctrl.js
Comment (mbayopanda) [reply client/src/modules/tags/tags.ctrl.js]: @jniles we have these three variants, and i am truggling a bit : 
```js
$ctrl.getTagColor = t => {
  return t ? { color : t.color } : null;
};

function addColorStyle(tag) {
  tag.style = { color : tag.color };
}

function addIconStyle(item) {
  item.iconStyle = { color : item.value };
  return item;
}
```
Finally, the `font-size` doesn't matter.

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [inline client/src/js/components/bhTagSelect/bhTagSelect.js:22]: This returns a promise, right?  Instead of using `$rootScope.$on()`, let's just do this:
```js
$ctrl.createUpdateTasModal = () => {
  Tags.createUpdateTagsModal()
    .finally(() => loadTags());
};
```

Would that work to remove the event-emitter pattern?

(you might also use a `.then()` instead - not sure what the return values of the tags modal is)

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [inline client/src/js/components/bhTagSelect/bhTagSelect.js]: Minor nitpick - if `$ctrl.tagUuids.length === 0`, then `identifiers.length === 0`.  Since you are actually using `identifiers` on this line, I think it cleaner to write:
```js
$ctrl.tags = identifiers.length ? tags.filter(t => identifiers.includes(t.uuid) === false) : tags; 
```

File: client/src/modules/tags/templates/name.cell.html
Comment (jniles) [reply client/src/modules/tags/templates/name.cell.html]: `GridRow` is used for clicking on dropdown menu items in the `actions` cells of tests.  What do you need to the `data-row` selector for here?  What are you checking/clicking on?

File: server/controllers/stock/index.js
Comment (jniles) [inline server/controllers/stock/index.js:535]: So, if I am reading this code correctly, if we have N lots in the depot, we will issue N + 1 SQL requests:
```js
// first SQL request (num SQL requests = 1)
const data = await core.getLotsDepot(null, params);


const rows = await Promise.all(data.map(async (line) => {
  // one SQL request per tag 
  const tags = await Lots.getLotTags(db.bid(line.uuid));
  line.tags = tags;
  return line;
}));

// Total SQL requests N + 1
```

I'm not sure something like a raspberry pi could handle this kind of a query.  Especially if no one uses tags, we'll be slowing down our queries for nothing.

I propose you either:
  1. Do something like `GROUP_CONCAT()` to get the tag information in a single SQL query.  This will be messy, but doable.
  2. Do two queries and stitch the results together in javascript.  You are already looping through each lot once, and it will save us from hammering our SQL server.  It would look something like this:
```js
const lotUuids = data.map(row => db.bid(row.uuid));
const tags = await db.exec('SELECT lot_uuid, tag.color /*, etc ... */ WHERE lt.lot_uuid IN (?)', [lotUuids]);


// make a lot_uuid -> tags map.
const tagMap = _.groupdBy(tags, 'lot_uuid');

data.forEach(lot => {
  lot.tags = tagMap[lot.uuid];
});

// Total SQL requests 2.
// Loops are O(N + M) where length(lots) is N and length(tags) is M.

```

File: test/end-to-end/tags/tags.page.js
Comment (jniles) [inline test/end-to-end/tags/tags.page.js:18]: 🤣

File: server/controllers/stock/index.js
Comment (jniles) [reply server/controllers/stock/index.js:535]: I suppose as an improvement, you can check if `tags.length === 0` and return early if that is the case.  It would save a for loop in the default case.

**Review by jniles — COMMENTED** [review]
I'm still tracking down the issues with the end to end tests - they take a long time to run!

Overall, this is really good.  I made a proposal about restructuring your SQL requests so that we don't shoot off a huge number of small queries.

I have also made a proposal on the use of the event emitter.  Can you see if that works?  I think it's a lot less magical than the `$on()` events and makes it easier for us to go to vue/angular2/react if we ever do that.

File: client/src/modules/tags/templates/name.cell.html
Comment (jniles) [reply client/src/modules/tags/templates/name.cell.html]: I actually think this is breaking the end to end tests.  When I remove it, I think the end to end tests pass. 👍

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (mbayopanda) [reply client/src/js/components/bhTagSelect/bhTagSelect.js:22]: Sorry, it seems like it doesn't work

File: test/end-to-end/tags/tags.spec.js
Comment (jniles) [reply test/end-to-end/tags/tags.spec.js]: Please remove these skips.

**Comment by mbayopanda** [comment]
@jniles can i get a review please

**Comment by jniles** [comment]
Why did you use grid action?  That is a step backwards. :/

**Comment by mbayopanda** [comment]
GridRow still doesn't pass the test

File: client/src/js/components/bhTagSelect/bhTagSelect.js
Comment (jniles) [reply client/src/js/components/bhTagSelect/bhTagSelect.js:22]: Turns out it doesn't return a promise.  This isn't a blocking change.  I'll open a pull request after this lands to change the behavior.

**Review by jniles — APPROVED** [review]
This works for me!  Thanks @mbayopanda. All comments have been addressed.

bors r+

**Pull request merged**
